diff --git a/iscsi_2.6.22_compat.h b/iscsi_2.6.22_compat.h
new file mode 100644
index 0000000..041837d
--- /dev/null
+++ b/iscsi_2.6.22_compat.h
@@ -0,0 +1,26 @@
+#include <linux/version.h>
+#include <scsi/scsi_cmnd.h>
+
+#ifndef ISCSI_2622_COMPAT_H
+#define ISCSI_2622_COMPAT_H
+
+#if LINUX_VERSION_CODE <= KERNEL_VERSION(2,6,21)
+#define netlink_kernel_create(uint, groups, input, cb_mutex, mod) \
+	netlink_kernel_create(uint, groups, input, mod)
+#endif
+
+#define scsi_sg_count(cmd) ((cmd)->use_sg)
+#define scsi_sglist(cmd) ((struct scatterlist *)(cmd)->request_buffer)
+#define scsi_bufflen(cmd) ((cmd)->request_bufflen)
+
+static inline void scsi_set_resid(struct scsi_cmnd *cmd, int resid)
+{
+	cmd->resid = resid;
+}
+
+static inline int scsi_get_resid(struct scsi_cmnd *cmd)
+{
+	return cmd->resid;
+}
+
+#endif
diff --git a/iscsi_tcp.h b/iscsi_tcp.h
index 57c2317..0b706a3 100644
--- a/iscsi_tcp.h
+++ b/iscsi_tcp.h
@@ -27,6 +27,7 @@
 #define ISCSI_SG_TABLESIZE		SG_ALL
 #define ISCSI_TCP_MAX_CMD_LEN		16
 
+#include "iscsi_2.6.22_compat.h"
 struct crypto_hash;
 struct socket;
 struct iscsi_tcp_conn;
diff --git a/libiscsi.h b/libiscsi.h
index 8328bc7..fbe14c0 100644
--- a/libiscsi.h
+++ b/libiscsi.h
@@ -30,6 +30,8 @@
 #include "iscsi_proto.h"
 #include "iscsi_if.h"
 
+#include "iscsi_2.6.22_compat.h"
+
 struct scsi_transport_template;
 struct scsi_device;
 struct Scsi_Host;
diff --git a/scsi_transport_iscsi.c b/scsi_transport_iscsi.c
index 0dbcedd..c737b43 100644
--- a/scsi_transport_iscsi.c
+++ b/scsi_transport_iscsi.c
@@ -29,6 +29,7 @@
 #include <scsi/scsi_transport.h>
 #include "scsi_transport_iscsi.h"
 #include "iscsi_if.h"
+#include "iscsi_2.6.22_compat.h"
 
 #define ISCSI_SESSION_ATTRS 16
 #define ISCSI_CONN_ATTRS 11
-- 
1.5.1.2

